content/function_calling.qmd

```diff
@@ -4,9 +4,146 @@ toc-title: '![](../imgs/cover.jpg){width=240px}<br> <h3>Function Calling</h3>'
 
 # Function Calling
 
+Function calling is a technique used in large language models (LLMs) and AI agents to enhance their capability to provide more accurate and relevant responses, especially when handling complex tasks or questions that require specialized knowledge or external data.
+
+We already got to know function calling in chapter 3 of this course. There, we introduced agents, that already came with the ability to call predefined functions. In this chapter, we will go back to the basics of function calling using LLMs. 
+
+
 ## Code generation and function calling
 
-## Data analysis
+The basic idea of function calling is to use an LLM to generate valid, executable code from the user input. That is, the user's input is sent to the LLM, together with a prompt, urging it to return structured output in a specific format. This output can then be taken by the agent and executed. For this to work properly, of course, the generated output *must* be valid code (in our case python code). There are two approaches for that:
+
+1. **Code generation**: Here, we ask the LLM to generate a complete python script from the user input. This approach has the advantage of being simple and straightforward, but it can be prone to errors if the LLM does not fully understand the task at hand or if it makes mistakes in its code generation.
```

Comment:
It can also pose a severe security issue because this approach hinges on running generated code on your machine

Reply:
included